### DISCUSSION TITLE
Implementation structure

### DISCUSSION THREAD

**Discussion by JannesBrands** [Open-Agriculture/AgIsoStack-rs#25] [General]
Hey guys, 
yesterday night I tried to get a general view of what is already existing / working.
I faced some mocks that are used but does not exist concretely and I did not know 
if there is a plan or how we want to structure the stack. So if there is not already another
idea, I would start with an idea of a strcuture:

Path
+ Datalink-3
+ Network-4
+ NetworkManagement-5
+ VirtualTerminalClient-06
+ ApplicationMessages-7
+ TractorECU-9
+ TaskControllerClient-10
+ DiagnosticsService-12
+ FileServerClient-13
+ SequenceControllerClient-14
+ Common

That would be my idea, even I am not sure  to maybe combine 3, 4, 5.
It is based on 2 other Isobus drivers that already exist and are used commercially. 
Even if this is now actually not a big idea cause its actually only derived from the standard. But I also see the advantage in this, namely the overview, since everyone who has worked with the standard will find their way around much more easily. If I specifically search for something in the driver and the part 11783-06 is divided into several folders like ObjectPool, VtCommunication... I think it is a bit harder to find the things.

I would be curious to read about your thoughts, that we have a discussion about it and are ready to start with an united idea.

@ad3154 
@GwnDaan 
@Thom-de-Jong 
@Notgnoshi

**Comment by JannesBrands** [comment]
Also I would suggest that we create some pullrequests as they are already existing and mark them like @GwnDaan did, with the label referring to the structure like TaskControllerClient-10 and maybe a reference to the ISO 11783 that describes and defines the feature. I think we need to get a little more organization into this for the community that contributes. Like in my case I don't got the time to work on it every night and maybe even with some breaks of 1 or 2 weeks and it's hard to get started again if you didn't track close the process. And anyway it would be helpful for the guys of Raven (or I don't know if there is already an internal planning process @ad3154 @Notgnoshi).

**Reply by ad3154** [reply]
I would be fine with us adding labels like that if you'd like. I think naturally some work will cross these boundaries, so the practicality of it may be not be 100% perfect but it will at least help identify the major components of issues and PRs.

Regarding the structure you proposed, I think some of those things probably make sense to break up by standard, I would be fine with that personally, though I'd like to hear others' opinions as well. The high level pieces (things like TC, VT, sequence control) especially seem like good candidates for such a structure.

I personally think Datalink, Network, and Network management will inevitably have to be combined, as they are very tightly integrated with each other and require a lot of coordination between them.

Lastly, apologies that progress has slowed a little. We're currently off of our innovation sprint at Raven, so this is a lower priority at the moment, and I have to focus on the C++ stack in my free time as well.

For this stack, I think we should really prioritize closing out/merging the open PRs so that we can keep moving forward with integrating the things at the Network layer with what we've already started for the Datalink layer with address claiming. Plus, we should keep integrating what we can recycle from @Thom-de-Jong, as he has a lot of good stuff that can get us a head start on VT and other high-level protocols.

**Reply by JannesBrands** [reply]
Yeah, I was thinking the same way about combining that network management. Maybe another idea would be splitting it between the CAN and J1939 level or something called "core" and "base".

**Reply by Notgnoshi** [reply]
> And anyway it would be helpful for the guys of Raven (or I don't know if there is already an internal planning process

We do not have an internal planning process for this. I think whatever we do, it should be public here on GitHub (i.e., Discussions, Issues, Projects, etc).

> I think we need to get a little more organization into this for the community that contributes. 

Agreed. @ad3154 and I both believe that the Open-Agriculture community is incredible, and we should definitely work to ensure community openness and stewardship. I think what @ad3154 has suggested (finishing open PRs, and building out the Network and Data Link layers) is appropriate to start with. I'm not a subject matter expert though, so I am intimidated by attempting this myself :)

> with the label referring to the structure like TaskControllerClient-10 and maybe a reference to the ISO 11783 that describes and defines the feature.

I like this!

**Reply by JannesBrands** [reply]
There is no reason to be intimidated :). Grab the standard and try what you can. The best feeling is when you can try it out on the field later. I think that's what makes the agriculture community, too. More and more young people are coming in who have grown up in the agricultural environment and are very visceral technically.

**Comment by GwnDaan** [comment]
Hee, I wanted to hop in here as well, sorry for being so late haha.

I saw you made nice initial suggestion in #26, but there are a few things I feel like would be clearer in another form:

1. I feel like `data_link` can still be separated from `network_management`. They can may still be interconnected, but having a dedicated folder for data transfer can help in getting it more organized. 
2. The 'part numbers' in the folder structure seem bit redundant and counter intuitive. I think only using the 'tags' for issues/PR's with the numbers is enough and keeps the naming more consistent throughout the Rust ecosystem. (I personally have rarely seen any numbers used in paths)
3. Don't separate client/server, they overlap a lot, like really a lot. So for example the object pool of a vt client should be found in the same directory as the vt server.
4. Have a dedicated 'hardware' folder (for everything hardware related, e.g. the driver implementations, or sensors or whatever there might be coming in the future)

My suggestion of the paths:

- hardware_interface 
- data_link
- network_management
- virtual_terminal
- application_messages
- tractor_ecu
- task_controller
- diagnostic_service
- file_server
- sequence_controller

Let me know your thoughts 😄

**Reply by JannesBrands** [reply]
Hey,
the differentiation between datalink and network was initially my idea too. We would have to agree on that again. As far as I remember @ad3154 would be for the merger of these. But that would be something to discuss again. 

I personally find the numbers of the standard as an addition helpful. I have also already seen in other commercial Isobus implementations. But yes, I agree with you on this point so far, except that there is no good alternative in this respect that directly understandably points to which part of the standard this software part is. 

Regarding the client server issue, as far as I know this is supposed to stay with a client implementation first. I think a server implementation would be cool too, but maybe in another moment or maybe in a separate project. 

The hardware part is still missing, I have not yet considered. There would be the question whether this should take place directly on level of the Isobus implementation, i.e. in the same folder, or outside.

There could be different solutions:

One gives only the interface on the basis of the Isobus implementation and we create separate projects that focus on specific hardware / architectures. 

Or as you have already suggested, or it is already implemented in the driver.

Where I am personally quite against is this implementation within the Isobus implementation, i.e. how currently the socketcan implementation is together with the datalink layer stuff in one folder. In my opinion this confuses a lot of things.

Conceivable options:
/src 
      /hal 
           /esp32 
           /stm32
           /socketcan
           /pcan
or 
for each hardware an own projcet like:
AgIsoStack-rs-esp32 
/src
      /hal
      /main.rs 

Is someone of you guys at Agritechnica in Germany?  
@Thom-de-Jong 
@GwnDaan 
@ad3154 
@Notgnoshi

**Reply by ad3154** [reply]
As a general thought, I don't really want us to get too bogged down in naming at this early stage. I think it's important to note that ISO11783 is a standard for communication, not a suggestion on how to structure software. Indicating to consumers of the library what files correspond to implementation of certain parts of the standard seems useful, I agree, but we should be ready to do whatever makes the software most robust.

> I personally have rarely seen any numbers used in paths

I agree, I think it's a bit unusual, but they also don't bother me too much if they provide value to people trying to use the library. I guess I'd be interested in hearing more opinions about it. Plus we can always change it - it's very early in development. 

> As far as I remember @ad3154 would be for the merger of these.

I don't feel too strongly either way. I guess I don't mind too much keeping data link split from network management, I just think they will be extremely closely connected in some ways. The only real discrete parts of the stack that I would foresee being in data link would be the transport protocol (TP.CM/BAM), ETP and a handful of enumerated values. Keeping then in their own folder seems fine enough.

Different topic: I'd rather keep the hardware integration pieces within a single project (this repo) if possible. It has worked very well in the C++ stack I think.

> Where I am personally quite against is this implementation within the Isobus implementation, i.e. how currently the socketcan implementation is together with the datalink layer stuff in one folder.

I think this is temporary to bootstrap something testable on a real bus so we can work on the data link layer. I wouldn't be too concerned. I do think that we should be clear that whatever folder name we use for the hardware stuff is clearly distanced from the standard itself if we're going to be naming things after the parts of ISO11783. Maybe this would be an argument for keeping the part numbers? I am not sure. `hardware` seems fine to me as Daan mentioned.

> Don't separate client/server, they overlap a lot, like really a lot. 

Yeah, agreed, numerous enumerated values, the object pool format, etc is all the same.

Just thinking out loud with most of this - good discussion 🙂

**Reply by GwnDaan** [reply]
> As a general thought, I don't really want us to get too bogged down in naming at this early stage. 

I agree. Even though renaming can be cumbersome, I think it will be inevitable at some point. Though a good discussion is never harmful :)

> I personally find the numbers of the standard as an addition helpful. I have also already seen in other commercial Isobus implementations. But yes, I agree with you on this point so far, except that there is no good alternative in this respect that directly understandably points to which part of the standard this software part is.

I see, though it still feels unnatural to me seeing numbers in paths hahah. My thought was that one most likely would know the name of the standard if they were to know the part number, but on the other hand, they not necessarily have to know the part number if they were to know the name. In the latter case it might make life more difficult when importing modules as one would have to explicitly write down the associated numbers, or rely on some form of auto-completion to do it for them. On top of that, IMO an ideal library shouldn't require one to know the standard to build some functional stuff, just some good examples to look at should be enough in most cases.

Another approach could be to have some sort of documentation with the structure and where parts of the standards can be found. Maybe a `README.md` file inside the `src` folder or something. 

With that said, I'm still fine with leaving the numbers in if others find it useful 🙂

Let's just agree on some format and see how it goes. It surely will be interesting reading back this discussion in the future, and seeing how it worked out haha

**Reply by GwnDaan** [reply]
> Is someone of you guys at Agritechnica in Germany?

@JannesBrands  Ohh yeah, I'll most likely be there 1 or 2 days, not sure which ones though. Are you going as well?

**Reply by JannesBrands** [reply]
Yep, Ill be there too for one or two days. Probably on monday and saturday.  
Lets interconnect to dont spam these topics full with off topic stuff.
Maybe it is possible to put the link into the description, as far you agree on that @ad3154 
I think this would be a good opportunity.
@Notgnoshi 
@GwnDaan 
@Thom-de-Jong 

[t.me/openagriculture](https://t.me/openagriculture)

**Reply by ad3154** [reply]
I guess Discord is my first choice for something like that if that's an option for you. Most of the other open source communities I follow seem to be on there, like [SerenityOS](https://github.com/SerenityOS/serenity) and it's convenient to be able to categorize chats (such as C++ stack versus Rust stack versus random other stuff).

But at the same time, if telegram is easiest for the community I'm willing to download it, haha.

https://discord.gg/FPkNqNHR4j

Either way, if we take some amount of discussion off-platform we'll want to remain vigilant about making sure issues and ideas end up on GitHub as the source of truth for what works needs doing and who's doing it.

**Reply by GwnDaan** [reply]
> it's convenient to be able to categorize chats (such as C++ stack versus Rust stack versus random other stuff)

Yeah, with telegram you can do that as well. I think telegram is a little bit more used inside the agriculture community but in the end I don't think it'll make a difference

**Reply by ad3154** [reply]
OK, I'll grab it then

**Reply by JannesBrands** [reply]
We need some opinions how we continue on this to maybe change some things of #26 and merge it. 

To give you my opinion:
Im open for @GwnDaan's suggestion:

> hardware_interface
data_link
network_management
virtual_terminal
application_messages
tractor_ecu
task_controller
diagnostic_service
file_server
sequence_controller


Now that some time has passed and I've gotten more involved with Rust, I would agree with @GwnDaan's suggestion due to the formatting matching Rust. 

Just give this post a thumbs up if you agree and I'll change it tomorrow so that the pull request can also be merged or let me know if you have any other idea.
@ad3154 
@Notgnoshi 
@GwnDaan 
@Thom-de-Jong
